parlai/agents/language_model/language_model.py

```diff
@@ -0,0 +1,480 @@
+        for p in self.model.parameters():
+            p.data.add_(-self.lr, p.grad.data)
+
+    def reset(self):
+        """Reset observation and episode_done."""
+        self.observation = None
+
+    def share(self):
+        """Share internal states between parent and child instances."""
+        shared = super().share()
+        shared['dict'] = self.dict
+        shared['NULL_IDX'] = self.NULL_IDX
+        shared['END_IDX'] = self.END_IDX
+        if self.opt.get('numthreads', 1) > 1:
+            shared['model'] = self.model
+            self.model.share_memory()
+            shared['states'] = self.states
+        return shared
+
+    def observe(self, observation):
+        """Save observation for act.
+        If multiple observations are from the same episode, concatenate them.
+        """
+        #shallow copy observation (deep copy can be expensive)
+        obs = observation.copy()
+        seq_len = self.opt['seq_len']
+        is_training = True
+        if 'eval_labels' in obs:
+            is_training = False
+
+        if is_training:
+            if 'text' in obs:
+                vec = self.parse(obs['text'])
+                vec.append(self.END_IDX)
+                self.next_observe += vec
+            if 'labels' in obs:
+                vec = self.parse(obs['labels'][0])
+                vec.append(self.END_IDX)
+                self.next_observe += vec
+            if len(self.next_observe) < (seq_len + 1):
+                # not enough to return to make a batch
+                # we handle this case in batchify
+                # labels indicates that we are training
+                self.observation = {'labels': ''}
+                return self.observation
+            else:
+                vecs_to_return = []
+                total = len(self.next_observe) // (seq_len + 1)
+                for _ in range(total):
+                    observe = self.next_observe[:(seq_len + 1)]
+                    self.next_observe = self.next_observe[(seq_len + 1):]
+                    vecs_to_return += observe
+                dict_to_return = {'text': '', 'labels': '', 'text2vec': vecs_to_return}
+                self.observation = dict_to_return
+                return dict_to_return
+        else:
+            if 'text' in obs:
+                obs['text'] += ' __END__'
+            self.observation = obs
+            return obs
+
+    def repackage_hidden(self, h):
+        """Wraps hidden states in new Variables, to detach them from their history."""
+        if type(h) == Variable:
+            return Variable(h.data)
+        else:
+            return tuple(self.repackage_hidden(v) for v in h)
+
+    def get_target_loss(self, data, hidden, targets, y_lens):
+        """Calculates the loss with respect to the targets, token by token,
+           where each output token is conditioned on either the input or the
+           previous target token.
+        """
```

Comment:
should be handled by nn.CrossEntropyLoss when padding_idx=0 and size_average=True

(unless I misunderstand what the below could is supposed to do)